### PR TITLE
[DEVHAS-192] Integrate devfile 2.2.0 outerloop

### PR DESCRIPTION
**Pull request by maysunfaisal**
Signed-off-by: Maysun J Faisal <maysunaneek@gmail.com>

### What does this PR do?:

This PR integrates devfile 2.2.0 and parses the outerloop devfile information ie; the inline/YAML content from a devfile kubernetes component. It also generates a deployment template for `Component` that uses image or dockerfile rather than a git/devfile repo.

If Kubernetes resource ie; either deployment, service or route is not provided, then one will generated for you. This logic is in PR https://github.com/redhat-developer/gitops-generator/pull/34. Once https://github.com/redhat-developer/gitops-generator/pull/34 is merged, this PR will consume the latest GitOps library.

### Which issue(s)/story(ies) does this PR fixes:

https://issues.redhat.com/browse/DEVHAS-192

### PR acceptance criteria:

- [x] Unit/Functional tests

  

- [ ] Documentation 

   

- [ ] Client Impact

  

### How to test changes / Special notes to the reviewer:
Create an `Application` and a `Component`, you should see that the GitOps YAMLs generated are based on the outerloop info from the devfile

### CODE REVIEW

File: controllers/applicationsnapshotenvironmentbinding_controller.go [hunk 1]
```diff
@@ -253,7 +253,7 @@ func (r *SnapshotEnvironmentBindingReconciler) Reconcile(ctx context.Context, re
 		// Retrieve the commit ID
 		var commitID string
 		repoPath := filepath.Join(tempDir, applicationName)
-		if commitID, err = gitopsgen.GetCommitIDFromRepo(r.AppFS, r.Executor, repoPath); err != nil {
+		if commitID, err = gitopsgen.GetCommitIDFromRepo(r.AppFS, repoPath); err != nil {
```
Comment (johnmcollier) [inline controllers/applicationsnapshotenvironmentbinding_controller.go:256]: Can you rebase? We're no longer using this function to get the commit id.

File: controllers/applicationsnapshotenvironmentbinding_controller.go [hunk 2]
```diff
@@ -49,10 +49,10 @@ import (
...
 type SnapshotEnvironmentBindingReconciler struct {
 	client.Client
-	Scheme   *runtime.Scheme
-	Log      logr.Logger
-	AppFS    afero.Afero
-	Executor gitopsgen.Executor
+	Scheme *runtime.Scheme
+	Log    logr.Logger
+	AppFS  afero.Afero
+	// Executor gitopsgen.Executor
```
Comment (johnmcollier) [inline controllers/applicationsnapshotenvironmentbinding_controller.go:55]: Why's the executor being removed from all of the GitOps functions?

**Review by johnmcollier — COMMENTED** [review]
Left a first pass of comments for what you had in draft before I leave on PTO.

File: controllers/applicationsnapshotenvironmentbinding_controller.go (hunk 2 above)
Comment (maysunfaisal) [reply controllers/applicationsnapshotenvironmentbinding_controller.go:55]: This was from the main branch, rebased and it should be gone

**Comment by maysunfaisal** [comment]
@yangcao77 @johnmcollier could you PTAL? I am currently updating the functional tests but other than that, it should be ok to review. This is blocked by PR https://github.com/redhat-developer/gitops-generator/pull/34

**Comment by maysunfaisal** [comment]
@johnmcollier we should switch to `container-override` & `pod-override` attribute semantics in HAS like its done in devfile project https://github.com/devfile/library/blob/main/pkg/devfile/generator/utils.go#L754.

This way we can move away from the attribute constants used for over riding and doing a if block check for everything

File: controllers/component_controller.go [hunk 3]
```diff
@@ -438,10 +477,22 @@ func (r *ComponentReconciler) generateGitops(ctx context.Context, req ctrl.Reque
 		return fmt.Errorf("unable to create temp directory for GitOps resources due to error: %v", err)
 	}
 
+	deployAssociatedComponents, err := devfileParser.GetDeployComponents(compDevfileData)
```
Comment (johnmcollier) [inline controllers/component_controller.go:480]: Making a mental note here: I'll need to rebase after this has been merged to move this section into our gitops-generator go module/job.

**Review by johnmcollier — APPROVED** [review]

**Review by johnmcollier — APPROVED** [review]

**Pull request merged**